### PR TITLE
Move composer to main directory

### PR DESCRIPTION
**Pull request by alex40724**
This PR suggests to move composer.json/lock to the top directory. It also adds two script entries for running the tests. Since composer is currently our main build tool we can use the scripts to perform everyday tasks on the top level. No `cd libs/composer` necessary.

```
> composer dump-autoload
> composer test-php                 // runs all standard tests (without installed ILIAS)
> composer test-php-all            // runs all tests
```

This enables us to run other tools (e.g. for JS-Building or JS-Testing) all via composer scripts on the top level.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -1,14 +1,21 @@
 {
-	"name": "ilias/deps",
-	"description": "Dependencies of ILIAS",
+	"name": "ilias",
```
Comment (clue) [inline composer.json]: This should include a vendor and package name, so `ilias/ilias` would be recommended (https://getcomposer.org/doc/04-schema.md#name)

**Review by clue — COMMENTED** [review]
@alex40724 Moving this to the project root directory makes a whole lot of sense (both for common project tools and for newcomers to find their way around) :+1: 

The build is currently failing, it looks like the `.travis.yml` needs to be updated accordingly.

File: composer.json (hunk 1 above)
Comment (alex40724) [reply composer.json]: @clue Thanks for the hint. Fixed.

File: templates/default/delos.css [hunk 2]
```diff
@@ -18920,4 +18920,3 @@ table.mceToolbar td {
   border: 1px solid #ddd;
   border-radius: .25em;
 }
-/*# sourceMappingURL=delos.css.map */
```
Comment (mjansenDatabay) [inline templates/default/delos.css]: @alex40724 Maybe a bit fussy, but does this really belong to this PR ;)?

**Review by mjansenDatabay — APPROVED** [review]
LGTM!

**Review by klees — APPROVED** [review]
Hi @alex40724!

Thanks for this PR, I absolutely agree to move the composer.*-files to the top level. AFAIK we could move the location for the checked-out dependencies (`vendor` directory) to the top level as well. What do you think?

Regarding composer as build tool: I am not sure we really want to use composer that way in the future. We are in fact collecting quite a pile of scripts and automatisms for builds that deserve a common interface. But IMO a specialized tool would be better for the job and intended to propose `make` some time soon. Any opinion here?

Best regards!

File: templates/default/delos.css (hunk 2 above)
Comment (alex40724) [reply templates/default/delos.css]: I will take care of this, before merging.

**Comment by alex40724** [comment]
@klees Having `vendor` dir on the top level would be the default. On the other hand the libs dir currently hides these tool specific directories from the top level dir and allows to separate the js from the php world. Personally I do not have a strong preference in this.

I agree that another build tool might be more appropriate. I just witnessed that developers are not able to run the tests properly, since the commands are complicated and have changed in the past.

Using composer this way as a temporary solution for shortcuts seems feasible to me, since this is the main tool every developer currently has to use.

Once we agree to move to another build tool, I happily help to migrate the shortcuts to the new tool.

**Comment by klees** [comment]
> I agree that another build tool might be more appropriate. I just witnessed that developers are not able to run the tests properly, since the commands are complicated and have changed in the past.

They could use also use `./CI/PHPUnit/run_tests.sh`. This also deals with additional parameters, so we could e.g. add `--filter UI` to run specific tests. Still, this seems to be partly unknown, which strengthens our impression that we need a common interface for the various scripts. Maybe it would be also a good idea to use that script for the shortcut via composer to have one implementation for properly running tests only?

**Comment by kergomard** [comment]
Generally I very much like the idea of moving composer to the top-level and I would also like to keep the dependencies encapsulated in the lib folder.
I think though that discussion on where we want to end up with this is an important one to have and that we should have it from a "non-programmer" perspective, too, as composer has become part of the installation process. As a programmer I very much like the idea of having a way to run all different tasks in the build-process from the root through a single interface, as an administrator I very much like to have an "installation tool" somewhere (and I very much dislike that in composer I have to remember the --no-dev switch) and maybe I would even like to have some pre-install task on the same interface.
I would like to avoid a two step process when it comes to the installation, i.e. we first move composer to the root and then decide we want to hide it somehow. Thus I suggest to have a specific workshop on this, encouraging administrators to take part in it, too.

**Comment by kergomard** [comment]
Sorry, fat finger!

**Comment by alex40724** [comment]
@swiniker Does this PR have to wait for your workshop or would you accept it as an interim improvement?

**Comment by kergomard** [comment]
I'm not going to stand in your way. I think it would be good though to put the discussion on a workplan to be finished before the release of ILIAS 7.

**Review by Amstutz — APPROVED** [review]
Thanks for the initiative. Looks good. Regarding the issue of the vendor folder, no strong opinion here, however, I do not like extremely crowded directories. The Top dir. is already a large bazar IMO we should propably do some clean up first before we shift the vendor dir and probably even further js deps there. As said, could also live with moving vendor now.

**Comment by alex40724** [comment]
If accepted I will modify the commands of this commit accordingly: https://github.com/ILIAS-eLearning/ILIAS/commit/f279670c4484be8bffa218fa50782af4cff2eea1

**Comment by alex40724** [comment]
I suggest to do this for 6.0 as well, not only trunk.

**Comment by matthiaskunkel** [comment]
Jour Fixe, 17 FEB 2020 : We highly appreciate this PR and accept it for trunk as well as for ILIAS 6.

**Comment by alex40724** [comment]
Merged manually.

**Pull request closed without merging**